exersices/Ex06/logistic_regression.py

```diff
@@ -0,0 +1,268 @@
+    tol : float (default=1e-5)
+        The gradient descent iterations will converge when the gradient
+        norm is less than this.
+    learning_rate : float (default=0.01)
+        The step-size for the gradient descent updates.
+    random_state : np.random.random_state or int or None (default=None)
+        A numpy random state object or a seed for a numpy random state object.
+    """
+
+    def __init__(
+        self, max_iter=1000, tol=1e-5, learning_rate=0.01, random_state=None
+    ):
+        """Initialise a logistic regression instance.
+        The ``__init__`` method of scikit-learn estimators should not do any
+        logic or input validation. This is all taken care of in the ``fit``
+        method.
+        Parameters
+        ----------
+        max_iter : int (default=1000)
+            Maximum number of gradient descent iterations to run.
+        tol : float (default=1e-5)
+            The gradient descent iterations will converge when the gradient
+            norm is less than this.
+        learning_rate : float (default=0.01)
+            The step-size for the gradient descent updates.
+        random_state : np.random.random_state or int or None (default=None)
+            A numpy random state object or a seed for a numpy random state
+            object.
+        """
+
+        self.max_iter = max_iter
+        self.tol = tol
+        self.learning_rate = learning_rate
+        self.random_state = random_state
+
+    def _has_converged(self, coef, X, y):
+        """Whether the gradient descent algorithm has converged.
+        Returns True if the norm of the gradient is smaller than ``self.tol``,
+        mathematically.
+
+        Parameters
+        ----------
+        coef : np.ndarray(shape=(r,))
+            The weight vector, :math:`\mathbf{w}^{(k)}`
+        X : np.ndarray(shape=(n, r))
+            The data matrix (aka design or measurement matrix)
+        y : np.ndarray(shape=(n,))
+            The true class labels for each data point.
+        Returns
+        -------
+        has_converged : bool
+            True if the convergence criteria above is met, False otherwise.
+        """
+        return np.linalg.norm(logistic_gradient(coef, X, y)) < self.tol
+
+    def _fit_gradient_descent(self, coef, X, y):
+        """Fit the logisitc regression model to the data given initial weights
+        Gradient descent works by iteratively applying the following update
+        rule
+
+        Parameters
+        ----------
+        coef : np.ndarray(shape=(r,))
+            The initial guess for the coefficient vector.
+            May be modified inplace by the method.
+        X : np.ndarray(shape=(n, r))
+            The data matrix
+        y : np.ndarray(shape=(n,))
+            The target vector
+        Returns
+        -------
+        coef : np.ndarray(shape=(n,))
+            The logistic regression weights
+        """
+        count = 0
+        while True:
```

Comment:
You also have to check for the `_has_converged` criteria here (for early stopping if the coef has converged before running all iterations).